src/ghcup/utils.rs

```diff
@@ -1,34 +1,177 @@
+use std::fmt;
+use std::fmt::{Debug, Display};
+use std::str::FromStr;
+
+use itertools::Itertools;
+use lazy_static::lazy_static;
 use reqwest::Client;
+use serde::{Deserialize, Serialize};
 
-use super::parser::CONFIG_VERSION;
 use crate::error::{Error, Result};
 
-pub async fn get_yaml_url<'a>(base_url: &'a str, client: &'a Client) -> Result<String> {
-    let version_matcher = regex::Regex::new("ghcupURL.*(?P<url>https://.*yaml)").unwrap();
+use super::GhcupRepoConfig;
+
+lazy_static! {
+    static ref YAML_CONFIG_PATTERN: regex::Regex =
+        regex::Regex::new(r"ghcup-(?P<ver>\d.\d.\d).yaml").unwrap();
+}
+
+macro_rules! t {
+    ($e: expr) => {
+        if let Ok(e) = $e {
+            e
+        } else {
+            return None;
+        }
+    };
+}
+
+// order is reverted to derive Ord ;)
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Version {
+    pub patch: usize,
+    pub minor: usize,
+    pub major: usize,
+}
+
+impl Version {
+    pub const fn new(major: usize, minor: usize, patch: usize) -> Self {
+        Self {
+            major,
+            minor,
+            patch,
+        }
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)
+    }
+}
+
+impl FromStr for Version {
+    type Err = ();
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        s.split('.')
+            .collect_tuple()
+            .and_then(|(major, minor, patch): (&str, &str, &str)| {
+                Some(Version::new(
+                    t!(major.parse()),
+                    t!(minor.parse()),
+                    t!(patch.parse()),
+                ))
+            })
+            .ok_or(())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct FileInfo {
+    id: String,
+    name: String,
+    path: String,
+}
 
-    let ghcup_version_module = client
-        .get(&format!("{}/lib/GHCup/Version.hs", base_url))
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TagInfo {
+    name: String,
+    commit: CommitInfo,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommitInfo {
+    id: String,
+}
+
+impl TagInfo {
+    pub fn id(&self) -> &str {
+        &self.commit.id
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ObjectInfo {
+    id: String,
+    name: String,
+    version: Version,
+}
+
+impl ObjectInfo {
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+    pub fn version(&self) -> Version {
+        self.version
+    }
+}
+
+pub async fn fetch_last_tag(client: &Client, config: &GhcupRepoConfig) -> Result<String> {
+    let req = client.get(format!(
+        "https://{}/api/v4/projects/{}/repository/tags",
+        config.host,
+        urlencoding::encode(&*config.repo)
```

Comment:
Why do we need to encode? Isn't the `{}` be replaced with something like `user/repo`?

Reply:
https://docs.gitlab.com/ee/api/index.html#namespaced-path-encoding